include/nbl/asset/ICPUBuffer.h

```diff
@@ -6,73 +6,91 @@
 
 #include <type_traits>
 
-#include "nbl/core/alloc/null_allocator.h"
-
 #include "nbl/asset/IBuffer.h"
 #include "nbl/asset/IAsset.h"
 #include "nbl/asset/IPreHashed.h"
 
 namespace nbl::asset
 {
 
+struct SICPUBufferCreationParams
+{
+    size_t size;
+    void* data = nullptr;
+    size_t alignment = _NBL_SIMD_ALIGNMENT;
+    std::pmr::memory_resource* memoryResource = nullptr;
+};
+
 //! One of CPU class-object representing an Asset
+//! TODO: remove, alloc can fail, should be a static create method instead!
 /**
     One of Assets used for storage of large arrays, so that storage can be decoupled
     from other objects such as meshbuffers, images, animations and shader source/bytecode.
 
     @see IAsset
 */
-class ICPUBuffer : public asset::IBuffer, public IPreHashed
+class ICPUBuffer final : public asset::IBuffer, public IPreHashed
 {
-    protected:
-        //! Non-allocating constructor for CCustormAllocatorCPUBuffer derivative
-        ICPUBuffer(size_t sizeInBytes, void* dat) : asset::IBuffer({ dat ? sizeInBytes : 0,EUF_TRANSFER_DST_BIT }), data(dat) {}
-
     public:
-        //! Constructor. TODO: remove, alloc can fail, should be a static create method instead!
-        /** @param sizeInBytes Size in bytes. If `dat` argument is present, it denotes size of data pointed by `dat`, otherwise - size of data to be allocated.
-        */
-        ICPUBuffer(size_t sizeInBytes) : asset::IBuffer({0,EUF_TRANSFER_DST_BIT})
-        {
-            data = _NBL_ALIGNED_MALLOC(sizeInBytes,_NBL_SIMD_ALIGNMENT);
-            if (!data) // FIXME: cannot fail like that, need factory `create` methods
-                return;
+        ICPUBuffer(size_t size, void* data, std::pmr::memory_resource* memoryResource, size_t alignment, bool adopt_memory) :

```

Comment:
if you have a `create` then the constructor must be private/protected (not public)

Also why not take the creation struct instead of each member of it individually ?

Reply:
Can't make it private/protected because `make_smart_refctd_ptr<ICPUBuffer>()` can't construct it

Reply:
because the way to do it is, from within `create`
```cpp
smart_refctd_ptr<ICPUBuffer>(new ICPUBuffer(std::move(params)),dont_grab);
```

thats how all the other classes with non-public constructions and public factory methods do it